linker/validate.go

```diff
@@ -83,17 +90,31 @@ func (r *result) validateExtension(fld protoreflect.FieldDescriptor, handler *re
 	return nil
 }
 
-func (r *result) validateJSONNamesInFile(handler *reporter.Handler) error {
-	for _, md := range r.FileDescriptorProto().GetMessageType() {
-		if err := r.validateJSONNamesInMessage(md, handler); err != nil {
-			return err
-		}
+func (r *result) validatePacked(fld protoreflect.FieldDescriptor, handler *reporter.Handler) error {
+	if xtd, ok := fld.(protoreflect.ExtensionTypeDescriptor); ok {
+		fld = xtd.Descriptor()
 	}
-	for _, ed := range r.FileDescriptorProto().GetEnumType() {
-		if err := r.validateJSONNamesInEnum(ed, handler); err != nil {
-			return err
+
+	fd := fld.(*fldDescriptor) //nolint:errcheck
+	if !fd.proto.GetOptions().GetPacked() {
+		// if packed isn't true, nothing to validate
+		return nil
+	}
+	if fd.proto.GetLabel() != descriptorpb.FieldDescriptorProto_LABEL_REPEATED {
+		file := r.FileNode()
+		pos := file.NodeInfo(r.FieldNode(fd.proto).FieldLabel()).Start()
+		err := handler.HandleErrorf(pos, "packed option is only allowed on repeated fields")
+		if err != nil {
+			return nil
```

Comment:
Should this return the error?

Reply:
It should, that's a good catch.

The reason it does not impact tests is because the handler (in above call to `HandleErrorf`) can choose to return nil, which tells the compiler to keep going. At the end, even if no ever is returned in this flow, it will see that an error was reported and return that. This is to support tools like `buf`, where they want the compiler to process as much as possible and report as many errors as possible instead of failing fast.

So the downside to this returning nil was that it was failing to short-circuit -- even though the handler returned an error, indicating the process should terminate, it would continue to do more wasted work.

I've fixed this line and also found another similar issue: in the caller of this function, when we get to the end of processing the file, even if no error was returned, we should still consider the file an erroneous one if any errors were reported. (The impact of that is rather subtle: but it's the difference between the file being included in "partial" results on failure or not -- since it is erroneous, it is not supported to be returned in partial results, but it would have been.)

Reply:
> So the downside to this returning nil was that it was failing to short-circuit

Ye, hence the comment. I just wasn't sure what the correct behaviour was in the end, whether we gather as much as possible or bail early.